[17.0][MIG] l10n_es_aeat_mod349: Migration to version 17.0

manuelregidor opened the pull request.

Commit:
[ADD] l10n_es_aeat, l10n_es_aeat_347, l10n_es_aeat_349: Adds a generic module for aeat models, 347 module was portedto v6.0 and adds new module to print AEAT model 349.

Commit:
[IMP+FIX] l10n_es_aeat, l10n_es_aeat_mod347, l10n_es_aeat_mod349: Pequeñas mejoras generales (vistas, traducciones, código) y corrección pequeño bug por un olvido en la adaptación de la v5 a la v6

albertca committed:
[FIX] l10n_es_aeat_mod349: 3 things:

* Only use the first four digits of fiscal year's code to determine fiscal year's number and also give a decent message to the user if it's not a number instead of simply crashing.
* Hacer opcional el campo Operation Key en las facturas, permitiendo que esté en blanco.
* Ser un poco menos rigurosos con que el año esté en el código del año fiscal. Atención: deberíamos añadir un campo con este valor en lugar de hacer este hack.

Commit:
[FIX+RFR] l10n_es_aeat_mod349: 2 things:

* FIX: Corregimos una serie de errores que había en las rectificaciones a la hora de obtener las declaraciones originales, si había alguna que cumplía los requisitos pero sin pertenecer a un modelo. Quizás se solucionara mejor con un on_delete='cascade'.
* RFR: Cambio de licencia de GPL3 a AGPL3 en el resto de módulos de localización española, con el consentimiento de todos los autores expresado vía lista de correo del equipo de localización openerp-spain@googlegroups.com

albertca committed:
[FIX] l10n_es_aeat_mod349: Several fixes

* Only use the first four digits of fiscal year's code to determine fiscal year's number and also give a decent message to the user if it's not a number instead of simply crashing.
* Resolve a problem with python isinstance which considers 'False' is an integer and which makes OpenERP create operation_key field as integer instead of char.
* Convert a return None to return False, because the XML-RPC protocol doesn't support returning None.

JordiBForgeFlow committed:
Corrección bug 1021631 por el cual la aplicación da el error "NameError: global name 'pooler' is not defined" al iniciar el calculo.

albertca committed:
l10n_es_aeat_mod349: Permitir seleccionar varios períodos en lugar de sólo uno y arrelgar un bug en la generación del modelo cuando el partner tiene el campo title lleno.

Commit:
[ADD] l10n_es_aeat_mod349: Añadimos dos nuevas claves de operación M y H que son respectivamente Entregas intracomunitarias con exención de tasas y Entregas intracomunitarias con exención de tasas por un representante fiscal oficial fueron publicadas el 26/11/2011 en el boe. Se puede ver el artículo en: http://planartus.blogspot.es/1309888740/

Commit:
Commit inicial de una rama para la versión 7.0. Todos los módulos fueron marcados como no instalables mientras no se compruebe que funcionan o migren, para poder usarlos o probarlos se debería poner el atributo installable de los ficheros __openerp__.py de cada módulo a True

acysos committed:
[FIX] l10n_es_aeat_mod349: string sin cerrar en l10n_es_aeat_mod349-pot

Commit:
Launchpad automatic translations update.

Commit:
[REF] cambiados los osv por orm.Model y orm.TransientModel, añadido en os import de objetos de OpenERP, el namespace openerp, refactorizado a PEP-8 y adaptacion vistas a formato 7

Commit:
Launchpad automatic translations update.

Commit:
[IMP] Adaptación de los import a formato v7/v8

Commit:
[IMP] l10n_es_aeat_mod349: Diversas mejoras y adaptación al nuevo l10n_es_aeat.

* Traducciones.
* Vista del wizard BOE
* Vistas de facturas mejoradas
* Icono
* Asignar una clave a las facturas ya existentes.

Commit:
[FIX] Adaptación a v8 de account_balance_reporting, l10n_es_aeat_mod349 y l10n_es_payment_order

Commit:
Launchpad automatic translations update.

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Con nueva API

Commit:
[IMP] unificar campos en el módulo base + migrar a la nueva api.

omar7r committed:
[FIX] l10n_es_aeat_mod349: Se corrige un problema al calcular las facturas rectificativas a declarar.

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod349: Uso incorrecto de ondelete en definición de o2m
[FIX] l10n_es_aeat_mod349: Añadida restricción que impide eliminar facturas que aparezcan en registros de detalle

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349: 3 things:

* IMP: Fiscal position onchange in new API + speed imp + refactoring
* FIX: l10n_es_aeat_mod349: Remove check of names

  There's no need of a check about the name validity, because l10n_es_aeat
  already transform all the exported strings to valid ASCII characters
  (though unidecode), so I remove it entirely this check.

* FIX: Use first period year in export

  Relying in fiscal year code for getting the fiscal year is not safe,
  because you can use that code for defining your journal sequences (for
  example, with the module account_auto_fy_sequence).

  We can't use also the year of the fiscal year, because you can have
  split fiscal years (i. e. like the academic courses). What we uses then
  is the year of the first period of the declaration, as there's no
  legal possibility of having mixed periods from different years.

oca-transbot committed:
OCA Transbot updated translations from Transifex

antespi committed:
[FIX] l10n_es_aeat_mod349: Name field in 349 conflict + add _rec_name

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Don't depend on fiscal year code for the export

Roodin committed:
[FIX]. l10n_es_aeat_mod349 . Recuperar funcionalidad perdida para poder calcular el 349 en períodos distintos de los fiscales

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Search only one refund detail record

angelmoya committed:
[MIG] l10n_es_aeat_mod349

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349

* Pylint
* Use new API

[skip ci]

luismontalba committed:
[IMP] l10n_es_aeat_mod349: Based on account moves

JordiBForgeFlow committed:
refactoring de l10n_es_aeat_mod349. Mejoras incluídas:
* Mejora del rendimiento.
* Apuntes contables como fuente de datos. El motor del informe se construye a partir de apuntes
  contables, y la información de facturas se consulta cuando se necesita. De este modo aseguramos que
  los resultados serán consistentes con los otros modelos de AEAT, que también toman como fuente de
  datos los apuntes contables. Así evitamos posibles inconsistencias.
* Nuevo informe Qweb, que sustituye al antiguo RML
* Correcciones para exporación a formato BOE. Permite importar el modelo a la web de hacienda.
* Permite borrar operaciones que no se quiere declarar. El usuario puede eliminar las operaciones que
  no desee incluir en la presentación. Esto será especialmente útil en el caso de presentaciones
  complementarias.
* En declaraciones normales y complementarias se excluyen automáticamente operaciones ya
  presentadas anteriormente.
* Script de migración para tener en cuenta los cambios en el modelo de datos.

JordiBForgeFlow committed:
[MIG] l10n_es_aeat_mod349: Migration to 10.0

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349: Several fixes + imps to new approach

* Save in account.move.line directly the 349 export key via a selection field, so you are not bound to a map line record instead of mapping changes.
* Field renamings for better understanding.
* Map data with command `4` instead of `6` for allowing custom mappings.
* Added constraint for avoiding duplicated mapping lines with the same operation key.
* Fix tests asserts and code
* Fix refund records computation
* Complete behavior for imported move lines declarations
* Switch certain fields to computed ones, for avoiding extra code that can be handled easily with ORM, but without performance penalty thanks to norecompute ORM option.
* Remove useless check for invalid characters, as export now removes them

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 11.0

Commit:
[FIX] l10n_es_aeat_mod349: Don't repeat refund details

carlosdauden committed:
[FIX] l10n_es_aeat_mod349: 2 things:

* Raise error with amount zero records (invoices+refunds=0.0)
* l10n_es_aeat_mod349: Missed by name param assignation precision_rounding

oca-travis committed:
[UPD] Update l10n_es_aeat_mod349.pot

JordiBForgeFlow committed:
l10n_es_aeat_349: corrige error en el 349 cuando se generan apuntes contables
asociados a una factura que no apuntan directamente a una línea de factura.

aitorbouzas committed:
[FIX] l10n_es_aeat_mod349: Fix report
Fix readme

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Adjust views permissions + remove old authors

* If views are not restricted to invoicing users, salesman won't be able to see the invoices.
* Current code doesn't share anything with original one, in approach nor lines of code, so
  original authors shouldn't be in the manifest key. I keep the copyright any way for
  historical reasons.

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 12.0

* Standard procedure
* README by fragments

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Fix tax reference

Copy/paste operation left out one needed tax.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Proper string format

pedrobaeza committed:
[IMP] l10n_es_aeat: Add active field to export configs

This way, we don't initially see all the subconfigs for each model and we can concentrate
on main export configurations. This is got through `active` field default behaviour.
Archived config are still directly searchable through the subconfig selector in the
configuration line.

All models export configs have been adapted to this new approach.

Commit:
[12.0][FIX] l10n_es_aeat_mod349: report

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Handle correctly Greek partners

As the VAT number should start with 'EL', not the country code 'GR'.

oihane committed:
[FIX] l10n_es_aeat_mod349: take all lines related to refunded invoice

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (146 of 146 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod349/es/

Commit:
[FIX] l10n_es_aeat_mod349: refund groups

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Proper year and month grouping key when no refund link

Code was not adapted to new v12 date fields with Python objects.

Courtesy of Jesús Ramoneda

pedrobaeza committed:
[IMP] l10n_es_aeat: Add chatter + activities to AEAT models

* Include chatter (mail.thread)
* Include activities (mail.activity.mixin)
* Track AEAT report state change
* Include new admin user in AEAT group by default

oihane committed:
[FIX] l10n_es_aeat_mod349: month from python only with one digit period_type needs 2

jesusVMayor committed:
[FIX] l10n_es_aeat_mod349: Corregido cálculo de periodo en base al mes del apunte.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Remove children records when recovering report to draft

This way, we allow things like cancelling invoices that are included in the report.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Don't depends on tax description

Continuation of 611c24bc and later commits. This way, we can decouple tax label from
the matching, thanks to the XML-ID that Odoo gives to created taxes. We also
simplifies the code thanks to this.

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 82.5% (146 of 177 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod349/es/

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Refund record totals + details

**Steps to reproduce**

- Create an intra-community invoice with several lines in May.
- Create a 349 AEAT report for May and calculate it.
- Refund the invoice in June.
- Create a 349 report for June and calculate it.

**Current behavior**

You get double origin amount in the refund record.

**Expected behavior**

Get proper origin amount.

**Solution**

We have to pre-group refund details by origin invoice, and then process them in
batch for getting the total amount.

It includes regression test.

Commit:
[FIX] l10n_es_aeat_mod349: total refunds calculation

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: black, isort, prettier

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 13.0

- General procedure
- In report change type > statement_type + migration script
- Computation adapted
- Adapt account.invoice > account.move in logic + migration script
- Adapt tests

oca-travis committed:
[UPD] Update l10n_es_aeat_mod349.pot

miguel-S73 committed:
[IMP] l10n_es_aeat_mod349: black, isort, prettier

miguel-S73 committed:
[MIG] l10n_es_aeat_mod349: migration to 14.0

miguel-S73 committed:
[FIX] l10n_es_aeat_mod349: origin amount incorrect

The calculation that was being done until now in the field origin_amount is incorrect since the total of the lines of the invoice to rectify was being added. The correct way is to add all the lines of the invoices of the period where the lines of the invoice to rectify are. Also, we must keep in mind that if there are intermediate periods between the original period and the period where the rectification is taking place, it's necessary to check if there is any rectification of the original period in between these periods. This happens in this way because the right original_amount  will be the value of the total_operation_amount corresponding to the last period found in between the periods

This commit has been based from this OCA/l10n-spain#1579

oca-travis committed:
[UPD] Update l10n_es_aeat_mod349.pot

OCA-git-bot committed:
[UPD] README.rst

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Don't mangle VAT number

There's a VAT formatting method that adds country code if not present
in the partner vat. This method is not useful at all, as all European
VAT numbers should have country prefix for being validated in invoices,
so there's not too much sense to try to reformat it. There are also
cases where a big company has one country VAT number, but has fiscal
address in other country, resulting in an incorrect reformating.

Not even that, there are old style onchanges that are not ever
triggered, as the XML declaration disappeared or is not effective as the
view is readonly, becoming dead code.

This commit removes all this stuff, with no expected side effects.

pere-aquarian committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (177 of 177 strings)

Translation: l10n-spain-14.0/l10n-spain-14.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-14-0/l10n-spain-14-0-l10n_es_aeat_mod349/es/

ValentinVinagre committed:
[MIG] l10n_es_aeat_mod349: Migration to 15.0

cubells committed:
[FIX] l10n_es_aeat_mod349: rounded total operation amount

The sum of the untaxed amount of the refunds might be not rounded, which
leads to having a fake partner_refund_ok = False, although all refund
data is correct.

With this preventive rounding, we avoid the problem and have the awaited
total_operation_amount.

rafaelbn committed:
[15.0] Set the standar name for all AEAT modules

The same detail from here: https://github.com/OCA/l10n-spain/pull/2129#issue-1141983559

Just same name as other AEAT modules

OCA-git-bot committed:
l10n_es_aeat_mod349 15.0.1.0.2

MiquelRForgeFlow committed:
[UPD] l10n_es_aeat_mod349: Eficent -> ForgeFlow

juliap98 committed:
Added translation using Weblate (Catalan)

juliap98 committed:
Translated using Weblate (Catalan)

Currently translated at 84.1% (143 of 170 strings)

Translation: l10n-spain-15.0/l10n-spain-15.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-15-0/l10n-spain-15-0-l10n_es_aeat_mod349/ca/

juliap98 committed:
Translated using Weblate (Catalan)

Currently translated at 99.4% (169 of 170 strings)

Translation: l10n-spain-15.0/l10n-spain-15.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-15-0/l10n-spain-15-0-l10n_es_aeat_mod349/ca/

Noel000 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (170 of 170 strings)

Translation: l10n-spain-15.0/l10n-spain-15.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-15-0/l10n-spain-15-0-l10n_es_aeat_mod349/ca/

manuelregidor committed:
[IMP]l10n_es_aeat_mod349: Mostrar errores en líneas del 349

MiquelRForgeFlow committed:
[FIX] l10n_es_aeat_*: exceptions.Warning -> exceptions.UserError

OCA-git-bot committed:
l10n_es_aeat_mod349 15.0.1.0.3

OCA-git-bot committed:
l10n_es_aeat_mod349 15.0.1.1.0

yajo committed:
[FIX] l10n_es_aeat_mod349: broken form UI

Fixing subtree width not responsive. See https://github.com/odoo/odoo/pull/90946 for a similar situation where a similar fix was needed.

Took the chance to make more columns optional.

@moduon MT-1267

OCA-git-bot committed:
l10n_es_aeat_mod349 15.0.1.2.0

victoralmau committed:
[FIX] l10n_es_aeat_mod349: Set the date field to store=True to allow the records to be sorted.

OCA-git-bot committed:
l10n_es_aeat_mod349 15.0.1.3.0

FernandoRomera committed:
[16.0][MIG] l10n_es_aeat_mod349: Migration to 16.0

Commit:
[UPD] Update l10n_es_aeat_mod349.pot

OCA-git-bot committed:
[UPD] README.rst

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Include food taxes + missing intra-community ones

OCA-git-bot committed:
l10n_es_aeat_mod349 16.0.1.1.0

Commit:
[UPD] Update l10n_es_aeat_mod349.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/

Commit:
Translated using Weblate (Catalan)

Currently translated at 97.0% (165 of 170 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/ca/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (170 of 170 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/es/

Commit:
[UPD] Update l10n_es_aeat_mod349.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/

OCA-git-bot committed:
[UPD] README.rst

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 99.4% (169 of 170 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/es/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (170 of 170 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/es/

ljsalvatierra-factorlibre committed:
[FIX] l10n_es_aeat_mod349: refund intermediate periods

Use case:
* Invoice January 2023: 1.000€
* Credit note February 2023: -200€
* credit note March 2023: -200€

Without the fix:
* 349 January total amount 1000€ (correct)
* 349 February total rectified amount 800€ (correct) and original amount 1000€ (correct)
* 349 March total rectified amount 800€ (incorrect) and original amount 1000€ (incorrect)

With the fix:
* 349 January total amount 1000€ (correct)
* 349 February total rectified amount 800€ (correct) and original amount 1000€ (correct)
* 349 March total rectified amount 600€ (correct) and original amount 800€ (correct)

Commit:
[UPD] Update l10n_es_aeat_mod349.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/

Commit:
[UPD] Update l10n_es_aeat_mod349.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (170 of 170 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_aeat_mod349/es/

Commit:
[IMP] l10n_es_aeat_mod349: implement more error texts

OCA-git-bot committed:
[BOT] post-merge updates

HaraldPanten commented:
/ocabot migration l10n_es_aeat_mod349

manuelregidor commented:
@HaraldPanten @ValentinVinagre

HaraldPanten commented:
> @HaraldPanten @ValentinVinagre

Puedes revisar los tests? Gracias.

HaraldPanten reviewed (approved):
Revisión funcional OK en runboat.

@pedrobaeza go?

ValentinVinagre commented on l10n_es_aeat_mod349/models/account_move.py:
```diff
@@ -35,6 +34,9 @@ class AccountMoveLine(models.Model):
+        store=True,
     )
 
     @api.depends("tax_ids", "move_id.eu_triangular_deal")
```
sería necesario también indicar `tax_ids.l10n_es_aeat_349_operation_key`?

manuelregidor replied:
¿Te refieres al depends? Si es así, lo pensé y lo revisé. Vi que siempre se actualiza antes el campo l10n_es_aeat_349_operation_key en el impuesto y el recálculo solamente se lanza al cambiar el impuesto en la línea.

pedrobaeza replied:
No, no se debe incluir ese depends. Eso lanzaría un recálculo global de todas las facturas que tengan el impuesto en caso de que se cambiara la clave de percepción del mismo. Si eso ocurriera, no querríamos que se cambie retroactivamente.

ValentinVinagre commented on l10n_es_aeat_mod349/models/account_tax.py:
```diff
@@ -25,6 +25,7 @@ def _compute_l10n_es_aeat_349_operation_key(self):
             tax.l10n_es_aeat_349_operation_key = False
             for line in map_349:
-                if tax in tax.company_id.get_taxes_from_templates(line.tax_tmpl_ids):
+                taxes_ids = tax.company_id._get_tax_ids_from_xmlids(line.tax_xmlid_ids)
```
es necesaria esta separación? en un principio 'taxes_ids' solo se utiliza para la comprobación, no es necesario guardarlo en una variable.

manuelregidor replied:
Se almacena en una variable por legibilidad.

ValentinVinagre replied:
@pedrobaeza que opinas?

manuelregidor replied:
Los modelos 111 y 347 utilizan la función get_taxes_for_company de l10n.es.aeat.map.tax.line, modelo que en el 349 no se utiliza porque el mapeo es otro. Quizás una opción sería crear una función como get_taxes_for_company en el modelo aeat.349.map.line, que es el específico del modelo 349.

pedrobaeza replied:
Bueno, internamente, Python lo manipula de forma similar, así que no hay tema de rendimiento, y coincido que se lee mejor así.

ValentinVinagre commented on l10n_es_aeat_mod349/models/mod349.py:
```diff
@@ -329,10 +327,11 @@ def _account_move_line_domain(self, taxes):
         if not tax_templates:
             raise exceptions.UserError(_("No Tax Mapping was found"))
-        return self.get_taxes_from_templates(tax_templates)
+        taxes_ids = self.company_id._get_tax_ids_from_xmlids(tax_templates)
+        return self.env["account.tax"].search([("id", "in", taxes_ids)])
```
se podría devolver directamente.
`return self.env["account.tax"].search([
    ("id", "in", self.company_id._get_tax_ids_from_xmlids(tax_templates))
])`

manuelregidor replied:
De nuevo, se almacena en una variable por legibilidad.

ValentinVinagre commented on l10n_es_aeat_mod349/models/res_company.py:
```diff
@@ -0,0 +1,16 @@
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    def _get_tax_ids_from_xmlids(self, tax_templates):
```
porque se crea esta función? en caso de ser necesaria tendría que crearse en el 'l10n_es_aeat'.

manuelregidor replied:
Se llama desde dos puntos distintos y se ha creado para no repetir código. Sí, sería mejor ponerla en l10n_es_aeat, así que tenemos dos opciones:
- La quitamos y repetimos código
- Hacemos un IMP en l10n_es_aeat

ValentinVinagre replied:
yo haría un IMP ¿los demás modelos como lo hacen? @pedrobaeza que opinas?

manuelregidor replied:
Los modelos 111 y 347 utilizan la función get_taxes_for_company de l10n.es.aeat.map.tax.line, modelo que en el 349 no se utiliza porque el mapeo es otro. Quizás una opción sería crear una función como get_taxes_for_company en el modelo aeat.349.map.line, que es el específico del modelo 349.

pedrobaeza replied:
Sí, esto debe estar en el mapeado de la línea 349 como lo está en el mapeado regular para los otros modelos:

https://github.com/OCA/l10n-spain/blob/af67194b45931b5b318aaca41c3bbfcf46533cdd/l10n_es_aeat/models/l10n_es_aeat_map_tax_line.py#L56-L64

ValentinVinagre replied:
@manuelregidor este punto como ha quedado finalemente?

ValentinVinagre commented on l10n_es_aeat_mod349/security/ir.model.access.csv:
```diff
@@ -12,5 +12,5 @@
 "access_l10n_es_aeat_mod349_aeat_349_map_line_manager","AEAT 349 Model: Operation key Map - Account Manager","model_aeat_349_map_line","account.group_account_manager",1,1,1,1
-"access_account_tax_template_billing","Tax Template - Billing user","model_account_tax_template","account.group_account_invoice",1,0,0,0
-"access_account_tax_template_user","Tax Template - Account user","model_account_tax_template","account.group_account_user",1,0,0,0
+"access_account_tax_template_billing","Tax Template - Billing user","l10n_es_aeat.model_l10n_es_aeat_map_tax_line_tax","account.group_account_invoice",1,0,0,0
+"access_account_tax_template_user","Tax Template - Account user","l10n_es_aeat.model_l10n_es_aeat_map_tax_line_tax","account.group_account_user",1,0,0,0
```
https://github.com/OCA/l10n-spain/blob/17.0/l10n_es_aeat/security/ir.model.access.csv
estos permisos no tendrían que ir en el 'l10n_es_aeat'?

manuelregidor replied:
Se ha puesto en este módulo porque así estaba en v16 con el modelo account.tax.template, que con su desaparición se corresponde en este contexto con l10n.es.aeat.map.tax.line.tax

ValentinVinagre replied:
@pedrobaeza que opinas?

pedrobaeza replied:
Poned en su lugar como `compute_sudo=True` el campo `l10n_es_aeat_349_operation_key`

ValentinVinagre reviewed (commented):
algunos comentarios

HaraldPanten commented:
@pedrobaeza Podrías darnos feedback de las dudas que han surgido?

Gracias!

HaraldPanten commented:
@manuelregidor A tener en cuenta de cara la migración. Para añadir todos los commits: https://github.com/OCA/l10n-spain/pull/3331

A la espera de la resolución de las dudas planteadas.

yajo committed:
[FIX] l10n_es_aeat_mod349: permitir a los comerciales ver sus facturas

Regresión introducida en https://github.com/OCA/l10n-spain/commit/22d4155d64911f5a7b9d04eea71419cc9d8860f9#diff-cf203442de5cb2bdf3aa371f214d1651256a58baa63ac5659c844287d53b5121.

Fix https://github.com/OCA/l10n-spain/issues/3539.
@moduon MT-5826

HaraldPanten commented:
A tener en cuenta cuando llegue a su conclusión --> https://github.com/OCA/l10n-spain/pull/3540

Commit:
[IMP] l10n_es_aeat_mod349: functionality to get the country_code

HaraldPanten commented:
@manuelregidor A tener en cuenta --> https://github.com/OCA/l10n-spain/pull/3544

HaraldPanten commented:
@pedrobaeza Nos puedes despejar las dudas que surgieron? Básicamente:

https://github.com/OCA/l10n-spain/pull/3503#discussion_r1542722123
https://github.com/OCA/l10n-spain/pull/3503#discussion_r1542723541
https://github.com/OCA/l10n-spain/pull/3503#discussion_r1542721310

pedrobaeza commented on l10n_es_aeat_mod349/models/account_move.py:
```diff
@@ -35,6 +34,9 @@ class AccountMoveLine(models.Model):
         compute="_compute_l10n_es_aeat_349_operation_key",
+        # It needs to be stored so this field is searchable in
+        # _create_349_refund_records in l10n.es.aeat.mod349.report
+        store=True,
```
Esto no se puede hacer así sin más. La instalación o actualización del módulo en un sistema con miles de apuntes va a tardar mucho tiempo. Si se hace store, se debe añadir un pre_init_hook que lo calcule de forma eficiente por SQL o similar. Pero una mejor técnica puede ser implementar el `search` del campo. ¿Por qué ahora hace falta buscar por él? ¿No se puede hacer la búsqueda sin incluir este criterio y luego filtrar por el campo?

manuelregidor replied:
Se utiliza en esta búsqueda
https://github.com/OCA/l10n-spain/blob/27f956f5703011217496520a0882ef477fc9c1e8/l10n_es_aeat_mod349/models/mod349.py#L267-L273

Tenía pendiente proponer fix en v16 para esta cuestión una vez aprobado este PR. Quito el store=True e implemento el search.

HaraldPanten reviewed (changes requested):
@manuelregidor Creo que no se han tenido en cuenta los commits mencionados en la conversación, puedes añadirlos?

Por otra parte, en las facturas de cliente es visible la columna l10n_es_aeat_349_operation_key . Es necesario?

manuelregidor commented:
@HaraldPanten Los commits adicionales no se han tenido todavía en cuenta.

Con respecto a la visibilidad de la columna l10n_es_aeat_349_operation_key, hay cambios en el commit https://github.com/OCA/l10n-spain/pull/3540, que todavía no se ha añadido.

HaraldPanten commented:
@manuelregidor 

Podemos ir añadiendo, entonces:

#3331 (ya se fusionó)
#3540 (ya se fusionó)

Y quedamos a la espera de #3544 (que también deberemos añadir). A este creo que no le queda mucho, solo que los tests parece que están un poco perezosos, esta mañana...

Gracias.

pedrobaeza commented:
Parece que las GitHub actions no están arrancando en todo OCA. Pero https://www.githubstatus.com/ no reporta ningún fallo.

manuelregidor committed:
[IMP] l10n_es_aeat_mod349: black, isort, prettier

manuelregidor committed:
[17.0][MIG] l10n_es_aeat_mod349: Migration to 17.0

manuelregidor commented:
@HaraldPanten @pedrobaeza @ValentinVinagre  Ya están incluidos los 3 PRs que quedaban pendientes

HaraldPanten reviewed (approved):
Revisión funcional OK.

Probado que los errores añadidos aquí https://github.com/OCA/l10n-spain/pull/3331 también sean visibles en la nueva versión.

@pedrobaeza

pedrobaeza reviewed (approved):
/ocabot merge nobump
